### ISSUE TITLE
Suggest Type Specifications via TypER

### ISSUE THREAD

**Issue by robertoaloi** [erlang-ls/erlang_ls#415]
Typer is a Type annotator for Erlang programs included in OTP:

http://erlang.org/doc/man/typer.html

With some modifications, it could be used to suggest type specs to the user, when missing.

**Comment by luos** [comment]
I think integrating Gradualizer also would be worthwile: https://github.com/josefs/Gradualizer/

**Comment by jfacorro** [comment]
@luos How does Gradualizer compare to Dialyzer?

**Comment by luos** [comment]
Hi,

It can be done live, same way as the Compiler diagnostics work, and it would result in type errors. I made some tests with this language server, test implementation and it seemed to work however I did not have time to polish it, if you want I can maybe share a screenshot or code form it.

**Comment by jfacorro** [comment]
@luos Some screenshots would be nice. I was more curious regarding the difference between Gradualizer and **Dialyzer** though. Since Dialyzer is also doing type checking, what would be the advantage of using Gradualizer instead?

**Comment by robertoaloi** [comment]
There are a few talks about Gradualizer from the Code Beam and Elixir Conf
EU.

Eg
https://youtu.be/_u1NDuFsW2A

**Comment by robertoaloi** [comment]
Difference at 33.00

**Comment by michalmuskala** [comment]
While once complete gradualizer would probably be able to provide stronger guarantees than dialyzer, at this point it is unfinished and probably not ready for production use.

**Comment by jfacorro** [comment]
I tried it yesterday on `erlang_ls`. It found some inconsistencies between the declared types and the actual type being used, and also missing values for union types (e.g. `poi_kind()`). 

But there are things that could be improved, for example it claims that `lager:debug/2` doesn't exist, which it doesn't, but I guess it should be able to provide support for `parse_transforms`. 

Another thing which I'm not sure if it's the tool or something that I didn't know how to solve, is that when calling `compiler:file/2` from a function that returns the result from that call, the return type `compile:comp_ret/0` from `compile:file/2` is not exported, but Gradualizer complains even when the spec for the function specifies a subset of `compile:comp_ret/0`.

**Comment by robertoaloi** [comment]
My take on this is that Erlang LS should serve as a platform and users should be able to decide which tools they want to use for their use case. So, in this case, it should be possible for the user to specify something like:

```
diagnostics:
    dialyzer: false
    gradualizer: true
```

And test gradualizer, which I would keep as disabled by default. The challenge with this approach is not to get a bloated Erlang LS which includes all possible dependencies from the Erlang ecosystem.

**Comment by alanz** [comment]
> The challenge with this approach is not to get a bloated Erlang LS which includes all possible dependencies from the Erlang ecosystem.

If we get the right kind of plugin architecture, we can let users specify a dependency to include which then provides features, via implementing a standard behaviour or some such.

**Comment by erszcz** [comment]
I'd also be interested in seeing Gradualizer integrated.

I'm not a type theory expert, but my understanding is that Dialyzer uses a theory built from scratch for its purpose ("succcess typings"), whereas Gradualizer uses a more classic typing model imbued with "gradual typing" - a blend of static checks if specs are present, and "ordinary" dynamic Erlang otherwise. Gradual typing whitepapers are quite recent and this research has been used successfully in languages like TypeScript, typed Python, typed Scheme and Facebook's Hack.

AFAIU, Dialyzer uses flow analysis to tell if a program has any discrepancies, therefore requires use cases to give useful hints about implemented code. In other words, it checks if "user code" and "library code" work together, whether there are specs or not. It's more mature, but the errors are obscure.

Gradualizer, on the other hand, focuses more on making sure specs/types and the corresponding implementations match. Presence of "user code" is not required to verify "library code" - specs are. No specs on a function means no checks. It infers types from literals, checks operations on them and verifies against user-provided types and function spec. It never checks messages passing, but provides operators for annotating received messages for the programmer. My personal opinion is that it's a bit better for helping with the "types as documentation" approach. It has way better error reporting. It's also a bit less mature, meaning seldom crashes or false positives might pop up.

Due to the theoretical models, Dialyzer and Gradualizer discover different classes of errors, so they're not 100% interchangeable.

My experience is that [Gradualizer is way faster](https://github.com/erszcz/gradualizer-talk/blob/master/slides.md#gradualizer-vs-dialyzer-3), therefore it's possible to integrate it better into IDEs. [I've built a Gradualizer plugin for Vim that I use daily](https://github.com/erszcz/neomake/tree/erlang-gradualizer-wip), but I think having one tool regardless of the editor of choice is the way to go.

**Comment by erszcz** [comment]
@jfacorro 
> It [gradualizer] found some inconsistencies between the declared types and the actual type being used

I think it's especially useful for this purpose, which in turn improves the quality of documentation.

> [...] there are things that could be improved, for example it claims that lager:debug/2 doesn't exist, which it doesn't, but I guess it should be able to provide support for parse_transforms.

It's supported via the `--specs_override_dir` flag. This allows for providing extra `-spec` attributes for code which doesn't exist (generated via parse transforms) or which does not provide specs itself. With my Vim plugin it boils down to these extra settings in a project dir, so it's not a big deal:

```
# project local .vimrc
$ cat .vimrc
let g:neomake_erlang_gradualizer_specs_override_dir = '/Users/erszcz/work/my_project/mongooseim/priv/gradualizer_overrides'
$ cat priv/gradualizer_overrides/lager.specs.erl
-module(lager).

-spec lager:debug(string(), list()) -> ok.
-spec lager:info(string(), list()) -> ok.
-spec lager:warning(string(), list()) -> ok.
-spec lager:error(string(), list()) -> ok.
-spec lager:critical(string(), list()) -> ok.
```

Alternatively, Gradualizer can be run on `.beam` files, so generated could is not a problem anymore, but in this mode there's no way to check source code that's not syntactically valid, so it might be a problem for use inside an editor.

**Comment by erszcz** [comment]
@robertoaloi @alanz 

```
diagnostics:
    dialyzer: false
    gradualizer: true
```

> If we get the right kind of plugin architecture, we can let users specify a dependency to include which then provides features, via implementing a standard behaviour or some such.

Rebar3 has a programmatic API. Maybe it wouldn't be a bad idea to bundle it with erlang-ls and dynamically recompile the server based on the config? This would allow not to pull in all deps from the get-go, but only when they're needed. The erlang-ls glue code could be present all the time, but the actual deps would or would not depending on the config.

**Comment by robertoaloi** [comment]
@erszcz I think this should work happily with the new proposed architecture, based on the Build Server Protocol. See https://github.com/erlang-ls/erlang_ls/blob/master/specs/runtime_node.md

The plan is to release a `rebar3` plugin which implements support for the BSP. Then it should be possible to support both mechanisms without having to add them both in Erlang LS. The project will decide.

**Comment by robertoaloi** [comment]
The feature is now available in Erlang LS (see #841 ) and documented here: https://erlang-ls.github.io/articles/suggest-type-specs/

Feedback welcome. If we want to look at Gradualizer, I think that deserves its own ticket. Feel free to open one, but also consider contributing (some help will likely be required).
